File: tangula-android-http/src/main/java/com/tangula/android/http/HttpBase.java
package com.tangula.android.http;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import com.tangula.utils.JsonUtils;
import com.tangula.utils.function.BiConsumer;
import com.tangula.utils.function.Consumer;
import com.tangula.utils.function.Supplier;

import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.ArrayList;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import kotlin.jvm.functions.Function2;
import kotlin.jvm.functions.Function3;
import kotlin.jvm.functions.Function6;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.HttpUrl;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;


@SuppressWarnings({"unchecked","unused"})
public abstract class HttpBase {


    public static Supplier<String> USER_ID_SUPPLIER ;

    public static Consumer<String> FUNC_LOG_VERB ;

    public static Consumer<String> FUNC_SHOW_MESSAGE ;

    public static BiConsumer<String, Throwable> FUNC_LOG_WARN ;

    public static BiConsumer<String, Throwable> FUNC_LOG_ERROR ;

    public static MediaType MEDIA_JSON = MediaType.parse("application/json; charset=utf-8");

    public static String URL_PREFIX = "";

    public static Function2<String, Object, Request> METHOD_GET;

    public static Function2<String, Object, Request> METHOD_POST_JSON ;

    public static Function2<String, Object, Request> METHOD_PUT_JSON;

    public static Function2<String, Object, Request> METHOD_DELETE_JSON;

    public static Function2<Response, Call, Unit> METHOD_COMMON_HTTP_FAIL_RESP;


    public static Function2<IOException, Call, Unit> METHOD_COMMON_HTTP_FAIL ;

    public static Function3<BizResponse<Object>, Response, Call, Unit> METHOD_COMMON_BIZ_FAIL ;

    public static void loadImage(String imageId, final Consumer<Bitmap> cb) {
        String url = URL_PREFIX + "/viewImageController/view/" + imageId;

        HttpUrl.Builder url_fac = HttpUrl.get(url).newBuilder();

        Request.Builder builder = new Request.Builder().url(url_fac.build());
        builder.addHeader("auth", USER_ID_SUPPLIER.get());

        Request request = builder.get().build();

        OkHttpClient client = new OkHttpClient();

        client.newCall(
                request
        ).enqueue(
                new Callback() {
                    public void onFailure( @NotNull Call call, @NotNull IOException e) {
                            FUNC_LOG_ERROR.accept(e.getLocalizedMessage(), e);
                    }

                    public void onResponse( @NotNull Call call, @NotNull Response response) throws IOException {
                        if (response.isSuccessful()) {
                            ResponseBody body = response.body();
                            if (body != null) {
                                byte[] bytes = body.bytes();
                                try {
                                    cb.accept(BitmapFactory.decodeByteArray(bytes, 0, bytes.length));
                                } catch (Exception e) {
                                    FUNC_LOG_ERROR.accept("image $url response not a picture", e);
                                    cb.accept(null);
                                }
                            } else {
                                FUNC_LOG_WARN.accept("image $url response not have a body", null);
                                cb.accept(null);
                            }
                        } else {
                            FUNC_LOG_WARN.accept("image $url response not success", null);
                            cb.accept(null);
                        }
                    }
                }
        );

    }


    protected static <T, R> void getBizSucessA(final String url, final T param, final Class<R> respType, final Function1<R, Unit> onSuccess)

    {
        getBizHttpSucessA(url, param, respType, onSuccess, (rBizResponse, response, call) -> {
            METHOD_COMMON_BIZ_FAIL.invoke((BizResponse<Object>) rBizResponse, response, call);
            return Unit.INSTANCE;
        });
    }

    protected static <T, R> void postBizSucessA(final String url, final T param, final Class<R> respType, final Function1<R, Unit> onSuccess)

    {
        postBizHttpSucessA(url, param, respType, onSuccess, (rBizResponse, response, call) -> {
            METHOD_COMMON_BIZ_FAIL.invoke((BizResponse<Object>) rBizResponse, response, call);
            return Unit.INSTANCE;
        });
    }


    public static <T, R> void getBizListSucessA(String url, T param, final Class<R> itemType, final Consumer<PagingBody<R>> onSuccess) {
        Function1<PagingBody, Unit> onSuc = body -> {
            PagingBody res = new PagingBody();
            res.setPageIndex(body.getPageIndex());
            res.setPageSize(body.getPageSize());
            res.setTotal(body.getTotal());
            final ArrayList<R> items = new ArrayList();

            for (Object o : body.getItems()) {
                items.add(JsonUtils.fromJson(JsonUtils.toJson(o), itemType));
            }

            res.setItems(items);
            onSuccess.accept(res);

            return Unit.INSTANCE;
        };
        getBizHttpSucessA(url, param, PagingBody.class, onSuc, (rBizResponse, response, call) -> {
            METHOD_COMMON_BIZ_FAIL.invoke((BizResponse) rBizResponse, response, call);
            return Unit.INSTANCE;
        });
    }


    public static <T, R> void postBizListSucessA(String url, T param, final Class<R> itemType, final Consumer<PagingBody<R>> onSuccess) {
        Function1<PagingBody, Unit> onSuc = body -> {
            PagingBody res = new PagingBody();
            res.setPageIndex(body.getPageIndex());
            res.setPageSize(body.getPageSize());
            res.setTotal(body.getTotal());
            final ArrayList<R> items = new ArrayList();

            for (Object o : body.getItems()) {
                items.add(JsonUtils.fromJson(JsonUtils.toJson(o), itemType));
            }

            res.setItems(items);
            onSuccess.accept(res);

            return Unit.INSTANCE;
        };
        postBizHttpSucessA(url, param, PagingBody.class, onSuc, (rBizResponse, response, call) -> {
            METHOD_COMMON_BIZ_FAIL.invoke((BizResponse) rBizResponse, response, call);
            return Unit.INSTANCE;
        });
    }

    protected <T, T1> void putBizSucessA(final String url, final T param, final Class<T1> respType, final Function1<T1, Unit> onSuccess)

    {
        putBizHttpSucessA(url, param, respType, onSuccess, (rBizResponse, response, call) -> {
            METHOD_COMMON_BIZ_FAIL.invoke((BizResponse<Object>) rBizResponse, response, call);
            return Unit.INSTANCE;
        });
    }


    protected <T, TR> void deleteBizSucessA(final String url, final T param, final Class<TR> respType, final Function1<TR, Unit> onSuccess)

    {
        deleteBizHttpSucessA(url, param, respType, onSuccess, (rBizResponse, response, call) -> {
            METHOD_COMMON_BIZ_FAIL.invoke((BizResponse<Object>) rBizResponse, response, call);
            return Unit.INSTANCE;
        });
    }


    protected static <T, R> void getBizHttpSucessA(final String url, final T param, final Class<R> respType, final Function1<R, Unit> onSuccess, final Function3<BizResponse<R>, Response, Call, Unit> onBizFail)

    {
        getBizA(url, param, respType, onSuccess, onBizFail, METHOD_COMMON_HTTP_FAIL_RESP, METHOD_COMMON_HTTP_FAIL);
    }

    protected static <T, R> void postBizHttpSucessA(final String url, final T param, final Class<R> respType, final Function1<R, Unit> onSuccess, final Function3<BizResponse<R>, Response, Call, Unit> onBizFail)

    {
        postBizA(url, param, respType, onSuccess, onBizFail, METHOD_COMMON_HTTP_FAIL_RESP, METHOD_COMMON_HTTP_FAIL);
    }

    protected static <T, R> void putBizHttpSucessA(final String url, final T param, final Class<R> respType, final Function1<R, Unit> onSuccess, final Function3<BizResponse<R>, Response, Call, Unit> onBizFail)

    {
        putBizA(url, param, respType, onSuccess, onBizFail, METHOD_COMMON_HTTP_FAIL_RESP, METHOD_COMMON_HTTP_FAIL);
    }


    protected static <T, R> void deleteBizHttpSucessA(final String url, final T param, final Class<R> respType, final Function1<R, Unit> onSuccess, final Function3<BizResponse<R>, Response, Call, Unit> onBizFail)

    {
        deleteBizA(url, param, respType, onSuccess, onBizFail, METHOD_COMMON_HTTP_FAIL_RESP, METHOD_COMMON_HTTP_FAIL);
    }

    protected static <T, R> void getBizA(final String url, final T param, final Class<R> respType, final Function1<R, Unit> onSuccess, final Function3<BizResponse<R>, Response, Call, Unit> onBizFail,
                                         final Function2<Response, Call, Unit> onFailResponse, final Function2<IOException, Call, Unit> onHttpFail)

    {

        httpBizA((s, t, aClass, onSuccess2, onFailResponse2, onHttpFail2) -> {
            sendHttpJsonARespJson(METHOD_GET, s, t, aClass, onSuccess2, onFailResponse2, onHttpFail2);
            return Unit.INSTANCE;
        }, url, param, respType, onSuccess, onBizFail, onFailResponse, onHttpFail);
    }

    protected static <T, R> void postBizA(final String url, final T param, final Class<R> respType, final Function1<R, Unit> onSuccess, final Function3<BizResponse<R>, Response, Call, Unit> onBizFail,
                                          final Function2<Response, Call, Unit> onFailResponse, final Function2<IOException, Call, Unit> onHttpFail)

    {
        httpBizA((s, t, aClass, onSuccess2, onFailResponse2, onHttpFail2) -> {
            sendHttpJsonARespJson(METHOD_POST_JSON, s, t, aClass, onSuccess2, onFailResponse2, onHttpFail2);
            return Unit.INSTANCE;
        }, url, param, respType, onSuccess, onBizFail, onFailResponse, onHttpFail);

    }


    protected static <T, R> void putBizA(final String url, final T param, final Class<R> respType, final Function1<R, Unit> onSuccess, final Function3<BizResponse<R>, Response, Call, Unit> onBizFail,
                                         final Function2<Response, Call, Unit> onFailResponse, final Function2<IOException, Call, Unit> onHttpFail)

    {
        httpBizA((s, t, aClass, onSuccess2, onFailResponse2, onHttpFail2) -> {
            sendHttpJsonARespJson(METHOD_PUT_JSON, s, t, aClass, onSuccess2, onFailResponse2, onHttpFail2);
            return Unit.INSTANCE;
        }, url, param, respType, onSuccess, onBizFail, onFailResponse, onHttpFail);
    }


    protected static <T, R> void deleteBizA(final String url, final T param, final Class<R> respType, final Function1<R, Unit> onSuccess,
                                            final Function3<BizResponse<R>, Response, Call, Unit> onBizFail,
                                            final Function2<Response, Call, Unit> onFailResponse, final Function2<IOException, Call, Unit> onHttpFail)

    {
        httpBizA((s, t, aClass, onSuccess2, onFailResponse2, onHttpFail2) -> {
            sendHttpJsonARespJson(METHOD_DELETE_JSON, s, t, aClass, onSuccess2, onFailResponse2, onHttpFail2);
            return Unit.INSTANCE;
        }, url, param, respType, onSuccess, onBizFail, onFailResponse, onHttpFail);
    }

    private static <T, R> void httpBizA(
            Function6<
                    String,
                    T,
                    Class,
                    Function3<BizResponse<R>, Response, Call, Unit>,
                    Function2<Response, Call, Unit>,
                    Function2<IOException, Call, Unit>, Unit> func,
            String url, T param, final Class<R> respType, final Function1<R, Unit> onSuccess,
            final Function3<BizResponse<R>, Response, Call, Unit> onBizFail,
            final Function2<Response, Call, Unit> onFailResponse, final Function2<IOException, Call, Unit> onHttpFail
    )

    {

        Function3<BizResponse<R>, Response, Call, Unit> resultHandler = (biz, resp, call) -> {
            if (biz != null) {
                if (biz.getStatus() == 0) {
                    FUNC_LOG_VERB.accept("resp business body type:" + respType.getName());
                    if (StringUtils.equalsAny(respType.getName(), "String", "char", "byte", "short", "double", "float", "long", "int", "boolean", "java.lang.Boolean", "java.lang.Integer", "java.lang.Long", "java.lang.Float", "java.lang.Double", "java.lang.Short", "java.lang.Byte", "java.lang.Character", "java.lang.String")) {
                        FUNC_LOG_VERB.accept("resp business body: ${biz.body}");
                        onSuccess.invoke( biz.getBody());
                    } else {
                        String strBody = JsonUtils.toJson(biz.getBody());
                        FUNC_LOG_VERB.accept("resp business body:  $strBody");
                        R respBody = JsonUtils.fromJson(strBody, respType);
                        onSuccess.invoke(respBody);
                    }
                } else {
                    onBizFail.invoke(new BizResponse(biz.getStatus(), biz.getStatusText(), biz.getMessage(), null), resp, call);
                }
            } else {
                onFailResponse.invoke(resp, call);
            }
            return Unit.INSTANCE;
        };

        func.invoke(URL_PREFIX + url, param, BizResponse.class, resultHandler, onFailResponse, onHttpFail);
    }


    protected static <T, R> void sendHttpJsonARespJson(Function2<String, Object, Request> method,
                                                       String url,
                                                       final T param, final Class<R> respType,
                                                       final Function3<R, Response, Call, Unit> onSuccess,
                                                       final Function2<Response, Call, Unit> onFailResponse,
                                                       final Function2<IOException, Call, Unit> onHttpFail)

    {
        OkHttpClient client = new OkHttpClient();
        client.newCall(
                method.invoke(url, param)
        ).enqueue(
                new Callback() {
                    public void onFailure( @NotNull Call call,  @NotNull IOException e) {
                        onHttpFail.invoke(e, call);
                    }

                    public void onResponse(  @NotNull Call call,  @NotNull Response response) throws IOException {
                        if (response.isSuccessful() || (response.code() >= 200 && response.code() <= 299)) {
                            ResponseBody body = response.body();
                            if (body != null) {
                                String str = body.string();
                                FUNC_LOG_VERB.accept("http resp body:$str");
                                FUNC_LOG_VERB.accept("http resp body type:" + respType.getName());
                                R obj = JsonUtils.fromJson(str, respType);
                                onSuccess.invoke(obj, response, call);
                            } else {
                                onSuccess.invoke(null, response, call);
                            }
                        } else {
                            onFailResponse.invoke(response, call);
                        }
                    }
                }
        );
    }


    public static void init() {
        USER_ID_SUPPLIER = new DefaultUserIdSupplier();

        FUNC_LOG_VERB = new DefaultConsoleMessageLog();

        FUNC_SHOW_MESSAGE = new DefaultConsoleMessageLog();

        FUNC_LOG_WARN = new DefaultConsoleErrorMessageLog();

        FUNC_LOG_ERROR =new DefaultConsoleErrorMessageLog();


        METHOD_GET = (url, param) -> {

            HttpUrl.Builder b = HttpUrl.get(url).newBuilder();

            for (Method prop : (param != null ? param : new Object()).getClass().getDeclaredMethods()) {
                String name = prop.getName();
                if (StringUtils.startsWith(name, "get") && (prop.getModifiers() & Modifier.PUBLIC) == Modifier.PUBLIC) {
                    try {
                        Object value = prop.invoke(param);
                        if (value != null) {
                            name = name.substring(3);
                            String fc = new String(new char[]{name.charAt(0)});
                            name = name.replaceFirst(fc, fc.toLowerCase());
                            b.addQueryParameter(name, value.toString());
                        }
                    } catch (IllegalAccessException e) {
                        FUNC_LOG_ERROR.accept(e.getLocalizedMessage(), e);
                    } catch (InvocationTargetException e) {
                        FUNC_LOG_ERROR.accept(e.getLocalizedMessage(), e);
                    }

                }
            }

            Request.Builder req = new Request.Builder().url(b.build());
            req.addHeader("auth", USER_ID_SUPPLIER.get());
            return req.get().build();

        };

        METHOD_POST_JSON = (url, param) -> {
            Request.Builder req = new Request.Builder().url(url);
            req.addHeader("auth", USER_ID_SUPPLIER.get());
            RequestBody body = RequestBody.create(MEDIA_JSON, JsonUtils.toJson(param != null ? param : new Object()));
            return req.post(body).build();
        };

        METHOD_PUT_JSON = (url, param) -> {
            Request.Builder req = new Request.Builder().url(url);
            req.addHeader("auth", USER_ID_SUPPLIER.get());
            RequestBody body = RequestBody.create(MEDIA_JSON, JsonUtils.toJson(param != null ? param : new Object()));
            return req.put(body).build();
        };

        METHOD_DELETE_JSON = (url, param) -> {
            Request.Builder req = new Request.Builder().url(url);
            req.addHeader("auth", USER_ID_SUPPLIER.get());
            RequestBody body = RequestBody.create(MEDIA_JSON, JsonUtils.toJson(param != null ? param : new Object()));
            return req.delete(body).build();
        };

        METHOD_COMMON_HTTP_FAIL_RESP = (res, call) -> {
            Request req = call.request();
            FUNC_LOG_WARN.accept("wrong response, req:" + req.method() + " " + req.url() + ", resp:" + res.toString() + ".", null);
            return Unit.INSTANCE;
        };


        METHOD_COMMON_HTTP_FAIL = (e, call) -> {
            Request req = call.request();
            FUNC_LOG_ERROR.accept("request fail, req:" + req.method() + " " + req.url() + ".", e);
            return Unit.INSTANCE;
        };

        METHOD_COMMON_BIZ_FAIL = (objectBizResponse, resp, call) -> {
            Request req = call.request();
            FUNC_LOG_ERROR.accept("business fail, req:" + req.method() + " " + req.url() + ", resp:" + resp.toString() + ".", null);
            return Unit.INSTANCE;
        };
    }

}

File: tangula-android-http/src/main/java/com/tangula/android/http/HttpBase.kt
package com.tangula.android.http

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.tangula.android.utils.UiThreadUtils
import com.tangula.utils.JsonUtils
import com.tangula.utils.function.BiConsumer
import com.tangula.utils.function.Consumer
import com.tangula.utils.function.Supplier
import okhttp3.*
import org.apache.commons.lang3.StringUtils
import java.io.IOException
import java.lang.Exception
import java.lang.reflect.Modifier


/**
 * Http工具类.
 * <p>使用这个类时，请继承这个类.类中的方法都是<code>protected</code>.这意味着必须对本类进行业务封装，否则不能使用其提供的Http访问接口</p>
 */
abstract class HttpBase {


    @Suppress("UNCHECKED_CAST")
    companion object {

        @JvmStatic
        var USER_ID_SUPPLIER: Supplier<String> = Supplier { "" }

        var FUNC_LOG_VERB:Consumer<String> = Consumer { it->
            println("[tag:http] $it")
        }

        var FUNC_LOG_WARN: BiConsumer<String, Throwable?> = BiConsumer { msg, e->
            println("[tag:http] $msg")
            e?.printStackTrace()
        }

        var FUNC_LOG_ERROR: BiConsumer<String, Throwable?> = BiConsumer { msg, e->
            println("[tag:http] $msg")
            e?.printStackTrace()
        }

        var FUNC_SHOW_MESSAGE:Consumer<String> = Consumer { it->
            println("[tag:http] $it")
        }

        @JvmStatic
        val MEDIA_JSON = MediaType.parse("application/json; charset=utf-8")

        @JvmStatic
        var  URL_PREFIX:String = ""

        @JvmStatic
        val METHOD_GET = { url: String, param: Any? ->
            val b = HttpUrl.get(url).newBuilder()

            for (prop in (param ?: object {}).javaClass.declaredMethods) {
                var name = prop.name
                if (StringUtils.startsWith(prop.name, "get") && prop.modifiers and Modifier.PUBLIC > 0) {
                    val value: Any? = prop.invoke(param)
                    if (value != null) {
                        name = name.drop(3)
                        name = name.replace(name[0], name[0].toLowerCase(), false)
                        b.addQueryParameter(name, value.toString())
                    }
                }
            }

            val req = Request.Builder().url(b.build())
            req.addHeader("auth", USER_ID_SUPPLIER.get())
            req.get().build()
        }

        @JvmStatic
        val METHOD_POST_JSON = { url: String, param: Any? ->
            val req = Request.Builder().url(url)
            req.addHeader("auth", USER_ID_SUPPLIER.get())
            val body = RequestBody.create(MEDIA_JSON, JsonUtils.toJson(param ?: object {}))
            req.post(body).build()
        }

        @JvmStatic
        val METHOD_PUT_JSON = { url: String, param: Any? ->
            val req = Request.Builder().url(url)
            req.addHeader("auth", USER_ID_SUPPLIER.get())
            val body = RequestBody.create(MEDIA_JSON, JsonUtils.toJson(param ?: object {}))
            req.put(body).build()
        }

        @JvmStatic
        val METHOD_DELETE_JSON = { url: String, param: Any? ->
            val req = Request.Builder().url(url)
            req.addHeader("auth", USER_ID_SUPPLIER.get())
            val body = RequestBody.create(MEDIA_JSON, JsonUtils.toJson(param ?: object {}))
            req.delete(body).build()
        }

        val METHOD_COMMON_HTTP_FAIL_RESP: (Response, Call) -> Unit = { res: Response, call: Call ->
            call.request().apply {
                FUNC_LOG_VERB.accept("wrong response, req:" + method() +" " + url() +", resp:"+ res.toString()+".")
            }
            FUNC_LOG_VERB.accept(res.toString())
        }

        val METHOD_COMMON_HTTP_FAIL: (IOException, Call) -> Unit = { ex: IOException, call: Call ->
            call.request().apply {
                FUNC_LOG_ERROR.accept("request fail, req:" + method() +" "+ url() +"", ex)
            }
        }

        val METHOD_COMMON_BIZ_FAIL: (BizResponse<Any>?, Response, Call) -> Unit = { biz: BizResponse<Any>?, resp: Response, call: Call ->
            call.request().apply {
                FUNC_LOG_VERB.accept("business fail, req:" + method() +" " + url() +", resp:"+ resp.toString()+".")
            }
            UiThreadUtils.runInUiThread(Runnable{
                FUNC_SHOW_MESSAGE.accept(biz?.message)
            })
        }

        @JvmStatic
        fun loadImage(imageId: String, cb: Consumer<Bitmap?>) {
            val url = "$URL_PREFIX/viewImageController/view/$imageId"

            val url_fac = HttpUrl.get(url).newBuilder()
            val builder = Request.Builder().url(url_fac.build())
            builder.addHeader("auth", USER_ID_SUPPLIER.get())
            val request = builder.get().build()

            val client = OkHttpClient()
            client.newCall(
                    request
            ).enqueue(
                    object : Callback {
                        override fun onFailure(call: Call, e: IOException) {
                            FUNC_LOG_ERROR.accept(e.localizedMessage, e)
                        }

                        @Throws(IOException::class)
                        override fun onResponse(call: Call, response: Response) {
                            if (response.isSuccessful) {
                                val body = response.body()
                                if (body != null) {
                                    val bytes = body.bytes()
                                    try {
                                        cb.accept(BitmapFactory.decodeByteArray(bytes, 0, bytes.size))
                                    } catch (e: Exception) {
                                        FUNC_LOG_ERROR.accept("image $url response not a picture", e)
                                        cb.accept(null)
                                    }
                                } else {
                                    FUNC_LOG_WARN.accept("image $url response not have a body",null)
                                    cb.accept(null)
                                }
                            } else {
                                FUNC_LOG_WARN.accept("image $url response not success",null)
                                cb.accept(null)
                            }
                        }
                    }
            )

        }

        @JvmStatic
        fun <T, R> getBizSucessA(url: String, param: T, respType: Class<R>, onSuccess: (R?) -> Unit) {
            getBizHttpSucessA(url, param, respType, onSuccess, METHOD_COMMON_BIZ_FAIL as (BizResponse<R>?, Response, Call) -> Unit)
        }

        @JvmStatic
        fun <T, R> postBizSucessA(url: String, param: T, respType: Class<R>, onSuccess: (R?) -> Unit) {
            postBizHttpSucessA(url, param, respType, onSuccess, METHOD_COMMON_BIZ_FAIL as (BizResponse<R>?, Response, Call) -> Unit)
        }

        /**
         * 列表数据
         */
        @JvmStatic
        fun <T, R> getBizListSucessA(url: String, param: T, itemType: Class<R>, onSuccess: Consumer<PagingBody<R>>) {
            getBizHttpSucessA(url, param, PagingBody::class.java, { body ->

                val res = PagingBody<R>().apply {
                    pageIndex = body!!.pageIndex
                    pageSize = body.pageSize
                    total = body.total
                }

                (body?.items as ArrayList<Object>).forEach {
                    val cur = JsonUtils.fromJson(JsonUtils.toJson(it), itemType)
                    (res.items as MutableList).add(cur)
                }

                onSuccess.accept(res)

            }, METHOD_COMMON_BIZ_FAIL as (BizResponse<PagingBody<*>>?, Response, Call) -> Unit)
        }


        /**
         * 列表数据
         */
        @JvmStatic
        fun <T, R> postBizListSucessA(url: String, param: T, itemType: Class<R>, onSuccess: Consumer<PagingBody<R>>) {
            postBizHttpSucessA(url, param, PagingBody::class.java, { body ->

                val res = PagingBody<R>().apply {
                    pageIndex = body!!.pageIndex
                    pageSize = body.pageSize
                    total = body.total
                }

                (body?.items as ArrayList<Object>).forEach {
                    val cur = JsonUtils.fromJson(JsonUtils.toJson(it), itemType)
                    (res.items as MutableList).add(cur)
                }

                onSuccess.accept(res)

            }, METHOD_COMMON_BIZ_FAIL as (BizResponse<PagingBody<*>>?, Response, Call) -> Unit)
        }


        @JvmStatic
        fun <T, R> putBizSucessA(url: String, param: T, respType: Class<R>, onSuccess: (R?) -> Unit) {
            putBizHttpSucessA(url, param, respType, onSuccess, METHOD_COMMON_BIZ_FAIL as (BizResponse<R>?, Response, Call) -> Unit)
        }

        @JvmStatic
        fun <T, R> deleteBizSucessA(url: String, param: T, respType: Class<R>, onSuccess: (R?) -> Unit) {
            deleteBizHttpSucessA(url, param, respType, onSuccess, METHOD_COMMON_BIZ_FAIL as (BizResponse<R>?, Response, Call) -> Unit)
        }


        @JvmStatic
        fun <T, R> getBizHttpSucessA(url: String, param: T, respType: Class<R>, onSuccess: (R?) -> Unit, onBizFail: (BizResponse<R>?, Response, Call) -> Unit) {
            getBizA(url, param, respType, onSuccess, onBizFail, METHOD_COMMON_HTTP_FAIL_RESP, METHOD_COMMON_HTTP_FAIL)
        }

        @JvmStatic
        fun <T, R> postBizHttpSucessA(uri: String, param: T, respType: Class<R>, onSuccess: (R?) -> Unit, onBizFail: (BizResponse<R>?, Response, Call) -> Unit) {
            postBizA(uri, param, respType, onSuccess, onBizFail, METHOD_COMMON_HTTP_FAIL_RESP, METHOD_COMMON_HTTP_FAIL)
        }

        @JvmStatic
        fun <T, R> putBizHttpSucessA(url: String, param: T, respType: Class<R>, onSuccess: (R?) -> Unit, onBizFail: (BizResponse<R>?, Response, Call) -> Unit) {
            putBizA(url, param, respType, onSuccess, onBizFail, METHOD_COMMON_HTTP_FAIL_RESP, METHOD_COMMON_HTTP_FAIL)
        }


        @JvmStatic
        fun <T, R> deleteBizHttpSucessA(url: String, param: T, respType: Class<R>, onSuccess: (R?) -> Unit, onBizFail: (BizResponse<R>?, Response, Call) -> Unit) {
            deleteBizA(url, param, respType, onSuccess, onBizFail, METHOD_COMMON_HTTP_FAIL_RESP, METHOD_COMMON_HTTP_FAIL)
        }


        @JvmStatic
        fun <T, R> getBizA(url: String, param: T, respType: Class<R>, onSuccess: (R?) -> Unit, onBizFail: (BizResponse<R>?, Response, Call) -> Unit,
                           onFailResponse: (Response, Call) -> Unit, onHttpFail: (ex: IOException, call: Call) -> Unit) {
            httpBizA(HttpBase.Companion::getA, url, param, respType, onSuccess, onBizFail, onFailResponse, onHttpFail)
        }

        @JvmStatic
        fun <T, R> postBizA(url: String, param: T, respType: Class<R>, onSuccess: (R?) -> Unit, onBizFail: (BizResponse<R>?, Response, Call) -> Unit,
                            onFailResponse: (Response, Call) -> Unit, onHttpFail: (ex: IOException, call: Call) -> Unit) {
            httpBizA(HttpBase.Companion::postJsonA, url, param, respType, onSuccess, onBizFail, onFailResponse, onHttpFail)
        }

        @JvmStatic
        fun <T, R> putBizA(url: String, param: T, respType: Class<R>, onSuccess: (R?) -> Unit, onBizFail: (BizResponse<R>?, Response, Call) -> Unit,
                           onFailResponse: (Response, Call) -> Unit, onHttpFail: (ex: IOException, call: Call) -> Unit) {
            httpBizA(HttpBase.Companion::putJsonA, url, param, respType, onSuccess, onBizFail, onFailResponse, onHttpFail)
        }


        @JvmStatic
        fun <T, R> deleteBizA(url: String, param: T, respType: Class<R>, onSuccess: (R?) -> Unit, onBizFail: (BizResponse<R>?, Response, Call) -> Unit,
                              onFailResponse: (Response, Call) -> Unit, onHttpFail: (ex: IOException, call: Call) -> Unit) {
            httpBizA(HttpBase.Companion::deleteJsonA, url, param, respType, onSuccess, onBizFail, onFailResponse, onHttpFail)
        }


        @JvmStatic
        private fun <T, R> httpBizA(func: (url: String, param: T, respType: Class<BizResponse<R>>, onSuccess: (BizResponse<R>?, Response, Call) -> Unit,
                                           onFailResponse: (Response, Call) -> Unit, onHttpFail: (IOException, Call) -> Unit) -> Unit,
                                    url: String, param: T, respType: Class<R>, onSuccess: (R?) -> Unit, onBizFail: (BizResponse<R>?, Response, Call) -> Unit,
                                    onFailResponse: (Response, Call) -> Unit, onHttpFail: (IOException, Call) -> Unit
        ) {
            return func(URL_PREFIX + url, param, BizResponse::class.java as Class<BizResponse<R>>, { biz: BizResponse<R>?, resp: Response, call: Call ->
                if (biz != null) {
                    if (biz.status == 0) {
                        FUNC_LOG_VERB.accept("biz.body" + JsonUtils.toJson(biz.body))  //TODO would remove
                        FUNC_LOG_VERB.accept( "biz.body type:" + respType.name) //TODO would remove
                        if (StringUtils.equalsAny(respType.name, "boolean", "java.lang.Boolean")) {
                            onSuccess(biz.body as R)
                        } else {
                            onSuccess(biz.body)
                        }
                    } else {
                        onBizFail(BizResponse<R>(biz.status, biz.statusText, biz.message, null), resp, call)
                    }
                } else {
                    onFailResponse(resp, call)
                }
            }, onFailResponse, onHttpFail)
        }

        @JvmStatic
        fun <T, R> getA(url: String, param: T, respType: Class<R>, onSuccess: (R?, Response, Call) -> Unit,
                        onFailResponse: (Response, Call) -> Unit, onHttpFail: (IOException, Call) -> Unit) {
            sendHttpJsonARespJson(METHOD_GET, url, param, respType, onSuccess, onFailResponse, onHttpFail)
        }

        @JvmStatic
        fun <T, R> postJsonA(url: String, param: T, respType: Class<R>, onSuccess: (R?, Response, Call) -> Unit,
                             onFailResponse: (Response, Call) -> Unit, onHttpFail: (ex: IOException, call: Call) -> Unit) {
            sendHttpJsonARespJson(METHOD_POST_JSON, url, param, respType, onSuccess, onFailResponse, onHttpFail)
        }

        @JvmStatic
        fun <T, R> putJsonA(url: String, param: T, respType: Class<R>, onSuccess: (R?, Response, Call) -> Unit,
                            onFailResponse: (Response, Call) -> Unit, onHttpFail: (ex: IOException, call: Call) -> Unit) {
            sendHttpJsonARespJson(METHOD_PUT_JSON, url, param, respType, onSuccess, onFailResponse, onHttpFail)
        }

        @JvmStatic
        fun <T, R> deleteJsonA(url: String, param: T, respType: Class<R>, onSuccess: (R?, Response, Call) -> Unit,
                               onFailResponse: (Response, Call) -> Unit, onHttpFail: (ex: IOException, call: Call) -> Unit) {
            sendHttpJsonARespJson(METHOD_DELETE_JSON, url, param, respType, onSuccess, onFailResponse, onHttpFail)
        }


        @JvmStatic
        fun <T, R> sendHttpJsonARespJson(method: (url: String, Any?) -> Request, url: String, param: T, respType: Class<R>, onSuccess: (R?, Response, Call) -> Unit,
                                         onFailResponse: (Response, Call) -> Unit, onHttpFail: (ex: IOException, call: Call) -> Unit) {
            val client = OkHttpClient()
            client.newCall(
                    method(url, param)
            ).enqueue(
                    object : Callback {
                        override fun onFailure(call: Call, e: IOException) {
                            onHttpFail(e, call)
                        }

                        @Throws(IOException::class)
                        override fun onResponse(call: Call, response: Response) {
                            if (response.isSuccessful||response.code() in 200 .. 299) {
                                val body = response.body()
                                if (body != null) {
                                    val str = body.string()
                                    FUNC_LOG_VERB.accept("http resp body:$str")
                                    FUNC_LOG_VERB.accept("http resp body type:" + respType.name)
                                    val obj = JsonUtils.fromJson(str, respType)
                                    onSuccess(obj, response, call)
                                } else {
                                    onSuccess(null, response, call)
                                }
                            } else {
                                onFailResponse(response, call)
                            }
                        }
                    }
            )
        }
    }


}
